sable_history/src/pg_history_service.rs

```diff
@@ -0,0 +1,278 @@
+use std::collections::HashMap;
+
+use chrono::{DateTime, NaiveDateTime, Utc};
+use diesel::dsl::sql;
+use diesel::prelude::*;
+use diesel_async::{AsyncPgConnection, RunQueryDsl};
+use futures::stream::{StreamExt, TryStreamExt};
+use tokio::sync::Mutex;
+use uuid::Uuid;
+
+use sable_network::prelude::*;
+
+use crate::schema::{channels, historic_users, messages};
+
+/// Implementation of [`HistoryService`] backed PostgreSQL
+pub struct PgHistoryService<'a> {
+    database_connection: &'a Mutex<AsyncPgConnection>,
+}
+
+impl<'a> PgHistoryService<'a> {
+    pub fn new(database_connection: &'a Mutex<AsyncPgConnection>) -> Self {
+        Self {
+            database_connection,
+        }
+    }
+}
+
+impl<'a> HistoryService for PgHistoryService<'a> {
+    async fn list_targets(
+        &self,
+        _user: UserId,
+        _after_ts: Option<i64>,
+        _before_ts: Option<i64>,
+        _limit: Option<usize>,
+    ) -> HashMap<TargetId, i64> {
+        // TODO: access control
+        // TODO: after_ts, before_ts, limit
+        match channels::dsl::channels
+            .select((
+                channels::dsl::id,
+                sql::<diesel::pg::sql_types::Uuid>(
+                    "SELECT MAX(id) FROM messages WHERE target_channel=channels.id",
+                ),
+            ))
+            .load_stream(&mut *self.database_connection.lock().await)
+            .await
+        {
+            Err(e) => {
+                tracing::error!("Could not get history channels: {e}");
+                return HashMap::new();
+            }
+            Ok(rows) => {
+                rows.map(|row| row.expect("Could not deserialize row"))
```

Comment:
Is panicking appropriate here or should we log an error and skip instead? We've got a lot of existing PoC code that panics all over the place but maybe it's time to make more efforts not to.

Reply:
Based on https://docs.rs/diesel/latest/diesel/result/enum.Error.html this happens when:

* there is a major logic error on our side,
* the database misbehaved, or
* the database schema does not match Rust's

It seems reasonable to crash either way.

Reply:
Heh, you're right. I removed the panics.